neo4j-jdbc/src/main/java/org/neo4j/jdbc/ConnectionImpl.java

```diff
@@ -667,8 +670,44 @@ Neo4jTransaction getTransaction(boolean autoCommitCheck) throws SQLException {
 	public void flushTranslationCache() {
 		synchronized (this) {
 			this.l2cache.flush();
-			this.sqlTranslator.resolve().flushCache();
+			this.sqlTranslator.resolve().forEach(SqlTranslator::flushCache);
 		}
 	}
 
+	static class TranslatorChain implements UnaryOperator<String> {
+
+		private final List<SqlTranslator> translators;
+
+		private final DatabaseMetaData metaData;
+
+		TranslatorChain(List<SqlTranslator> translators, DatabaseMetaData metaData) {
+			this.translators = translators;
+			this.metaData = metaData;
+		}
+
+		@Override
+		public String apply(String sql) {
+			if (this.translators.size() == 1) {
```

Comment:
Regarding readability / maintenance:
Is this special handling for size == 1 really needed?
Yes, the iterator is not invoked here and it _might_ have performance benefits but shouldn't take the JVM care of optimising this?

Reply:
Burned by this too many times wrt streams.